### PR TITLE
Add a HostFailureListener to notify client code if a node got disconnected

### LINKED ISSUE #21424: Transport client doesn't handle DNS -> IP binding changing
Today, when you instantiate the transport client, it resolves DNS -> IP on init and uses that IP forever.

If your DNS binding changes, the transport client will keep using the old IP, forcing you to know when to re-instantiate it to get the new DNS binding.

Maybe transport client should more frequently resolve, e.g. per request, or every N seconds?  This is possibly dangerous, though, in the face of flaky/slow DNS lookups, and it could add a sizable latency.

This is similar to https://github.com/elastic/elasticsearch/issues/16412 for unicast zen ping.

### PR DESCRIPTION
**Pull request by s1monw**
Today there is no way to get notified if a node is disconnected. Client code
must poll the TranportClient constantly to detect that a node is not connected
anymore in order to react and add new nodes or notify altering etc. For instance
if a hostname  gets resolved to an IP but that host is disconnected clients want
to reconnect by resolving the hostname again which is a common situation in cloud
environments.

Closes #21424

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java [hunk 1]
```diff
@@ -275,6 +287,16 @@ public void onFailure(Exception e) {
             }
         }
 
+        final DiscoveryNode getNode(int i) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java]: Very nice seeing this in one place now.

File: core/src/main/java/org/elasticsearch/client/transport/TransportClient.java [hunk 2]
```diff
@@ -164,7 +164,8 @@ private static ClientTemplate buildTemplate(Settings providedSettings, Settings
 
             Injector injector = modules.createInjector();
             final TransportClientNodesService nodesService =
-                new TransportClientNodesService(settings, transportService, threadPool);
+                new TransportClientNodesService(settings, transportService, threadPool, failureListner == null
+                    ? (t,e) -> {} :failureListner);
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/client/transport/TransportClient.java]: Nit: Do you mind fixing the spacing here: `(t,e) -> {} :failureListner` -> `(t, e) -> {} : failureListner`?

File: core/src/test/java/org/elasticsearch/client/transport/TransportClientNodesServiceTests.java [hunk 3]
```diff
@@ -101,7 +101,7 @@ public AsyncSender interceptSender(AsyncSender sender) {
             transportService.start();
             transportService.acceptIncomingRequests();
             transportClientNodesService =
-                    new TransportClientNodesService(settings, transportService, threadPool);
+                    new TransportClientNodesService(settings, transportService, threadPool, (a,b) -> {});
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/client/transport/TransportClientNodesServiceTests.java]: Nit: Do you mind fixing the spacing here: `(a,b)` -> `(a, b)`?

File: client/transport/src/main/java/org/elasticsearch/transport/client/PreBuiltTransportClient.java [hunk 4]
```diff
@@ -61,7 +61,10 @@ public PreBuiltTransportClient(Settings settings, Class<? extends Plugin>... plu
     }
 
     public PreBuiltTransportClient(Settings settings, Collection<Class<? extends Plugin>> plugins) {
-        super(settings, Settings.EMPTY, addPlugins(plugins, PRE_INSTALLED_PLUGINS));
+        this(settings, plugins, null);
+    }
+    public PreBuiltTransportClient(Settings settings, Collection<Class<? extends Plugin>> plugins, HostFailureListener hostFailureListener) {
```
Comment (jasontedor) [inline client/transport/src/main/java/org/elasticsearch/transport/client/PreBuiltTransportClient.java]: Nit: Can you add a newline here?

Comment (jasontedor) [inline client/transport/src/main/java/org/elasticsearch/transport/client/PreBuiltTransportClient.java]: Can you add a Javadoc to this constructor?

File: core/src/main/java/org/elasticsearch/client/transport/TransportClient.java [hunk 5]
```diff
@@ -332,4 +334,24 @@ public void close() {
...
+         */
+        void onNodeDisconnected(DiscoveryNode node, Exception ex);
+    }
+
+    // pkg private for testing
+    TransportClientNodesService getNodesService() {
+        return nodesService;
+    }
+
+
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/client/transport/TransportClient.java]: Nit: there's an extra newline here.

File: core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java [hunk 6]
```diff
@@ -275,6 +287,16 @@ public void onFailure(Exception e) {
...
+            return nodes.get((index + i) % nodes.size());
+        }
+
+        final void maybeNodeFailed(DiscoveryNode node, Exception ex) {
+            if (ex instanceof NodeDisconnectedException || ex instanceof NodeNotConnectedException) {
+                hostFailureListener.onNodeDisconnected(node, ex);
+            }
+        }
+
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java]: Nit: there's an extra newline here.

**Review by jasontedor — APPROVED** [review]
Very nice change, I like the tests. I left some nits, but you have an LGTM from me to push at your discretion.

**Comment by jasontedor** [comment]
test this please

**Comment by s1monw** [comment]
@jasontedor I pushed changed and cleaned some stuff up do you wanna take a quick look since there are some more changes but all of them mechanical..

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by s1monw** [comment]
test this please

**Comment by s1monw** [comment]
test this please

**Comment by s1monw** [comment]
I marked this as a bug since it can really be a showstopper when you use the transport client especially if you are in a cloud env.

**Comment by gangakrishh** [comment]
@s1monw can you provide example for this use case.

I'm running elasticsearch in aws, in the case of ip (elasticsearch EC2) re assign by aws , my application not able to connect ES

Client clode
`String host = internal-xy.us-east-1.elb.amazonaws.com
Settings settings = Settings.builder().put("cluster.name", "test").build();
 TransportClient client = new PreBuiltTransportClient(settings)
    			      .addTransportAddress(new InetSocketTransportAddress(InetAddress.getByName(host), 9300));`

**Comment by s1monw** [comment]
I can't tell you how you get this info from EC2 but you have to connect to their API and find the IP?

**Comment by gangakrishh** [comment]
@s1monw i didn't get you ..

is this approach recommended ?

https://gist.github.com/vaughnd/22fe0670a296ebcb436c

**Pull request merged**